### ISSUE TITLE
API Example

### ISSUE THREAD

**Issue by MMRandy** [lapwinglabs/x-ray#134]
Having some trouble trying to figure out how to handle situations when the HTML is not very cooperative.  Take this page for example: http://www.craigslist.org/about/sites

How would you go about creating a nested object with the following shape?

```
[
    {
        country: 'US',
        states: [
            {
                name: 'Alabama',
                cities: [
                    { name: 'auburn'},
                    { name: 'birmingham'}
                    ...
                ]
            },
            {
                name: 'Alaska',
                 cities: [
                    { name: 'anchorage / mat-su'},
                    ....
                ]
            },
            ...
        ]
    },
    {        
        country: 'Canada',
        states: [
            ....
        ]
        
    }
]
```

Seeing this done would definitely help me to understand more about about scoping selectors for sibling nodes that need to be part of the same object.  This is very easy with cheerio, thinking I just don't know enough about x-ray's API to be able to deal with siblings.

Thanks!

**Comment by dfdeagle47** [comment]
Here's how I would do it. I have no idea if this is clean or fast, but it works

Note: you might have to do more processing afterwards because states are actually countries for Europe, and cities are sometimes countries as well in this example.

```javascript
var Xray = require('x-ray');
var x = new Xray();

x('http://www.craigslist.org/about/sites', 'h1', [{
    regionCode: 'a@name',
    region: '',
    states: function($, cb) {
      return x($.next(), 'h4', [{
        name: '',
        cities: function($, cb2) {
          return x($.next(), 'li', [{
            name: 'a'
          }])(cb2);
        }
      }])(cb);
    }
  }])(function(err, obj) {
    console.log(err, obj);
  });
```

Basically, I use the fact that you can pass an arbitray function for properties. If you look at (index.js; lines 176-182):

```javascript
else if ('function' == typeof v) {
          return v($, function(err, obj) {
            if (err) return next(err);
            return next(null, obj);
          });
        }
```

You see that if your property is a function, it will be called using the parameters $ and a callback. $ represents the scope at that level. Basically, from what I understand, it comes from cheerio.load (line 166), or a 'sub'-scope of that. You can then use $.next to get the sibling or other methods on that object.

So, I just re-used the x-ray object, but with a different scope.

Let me know if you find something else because that's a pretty interesting problem since most pages are more or less garbage.

Note: you might have to use my fork for now https://github.com/dfdeagle47/x-ray because of Issue #135 because I tried with my fork, but not the original.

**Comment by dfdeagle47** [comment]
By the way, in this example, it would probably much easier to actually parse the response from [http://www.craigslist.org/about/areas.json](http://www.craigslist.org/about/areas.json) to get the information you need.

**Comment by MMRandy** [comment]
@dfdeagle47, sorry for tardy response...traveling right now.  But thanks a lot for this...was not aware that properties can resolve functions and accept cheerio type params to perform processing.  That's awesome.

I am in the process of comparing various scraping libs so the purpose was more related how the x-ray process compares to others vs the simplest way to transform the JSON.  But thanks again for your responses here, really helped :)

**Comment by zipzit** [comment]
Look again closely at Readme.md, the selection entitled **Scoping a selection**;  that section has multiple `x(...)` calls which work pretty well.  I too originally thought the "download the whole site and process it later" was a good idea, until I discovered I had a 500Kb download for each page on my screen scrape.  Ugh.  Clicking on that file would lock up atom.io, and yet I clicked on it every time.  With careful scoping I was able to get that scrape down to 18Kb, and it processes quicker too.    I think the nested `x(..` calls is pretty easy to use.
